File: src/spice2x/api/resources/python/spiceapi/rc4.py


def rc4_ksa(key):
    n = len(key)
    j = 0
    s_box = list(range(256))
    for i in range(256):
        j = (j + s_box[i] + key[i % n]) % 256
        s_box[i], s_box[j] = s_box[j], s_box[i]
    return s_box


def rc4_prga(s_box):
    i = 0
    j = 0
    while True:
        i = (i + 1) % 256
        j = (j + s_box[i]) % 256
        s_box[i], s_box[j] = s_box[j], s_box[i]
        yield s_box[(s_box[i] + s_box[j]) % 256]


def rc4(key):
    return rc4_prga(rc4_ksa(key))

File: src/spice2x/api/resources/cpp/spiceapi/rc4.cpp
#include "rc4.h"
#include <iterator>

spiceapi::RC4::RC4(uint8_t *key, size_t key_size) {

    // initialize S-BOX
    for (size_t i = 0; i < std::size(s_box); i++)
        s_box[i] = (uint8_t) i;

    // check key size
    if (!key_size)
        return;

    // KSA
    size_t j = 0;
    for (size_t i = 0; i < std::size(s_box); i++) {

        // update
        j = (j + s_box[i] + key[i % key_size]) % std::size(s_box);

        // swap
        auto tmp = s_box[i];
        s_box[i] = s_box[j];
        s_box[j] = tmp;
    }
}

void spiceapi::RC4::crypt(uint8_t *data, size_t size) {

    // iterate all bytes
    for (size_t pos = 0; pos < size; pos++) {

        // update
        a = (a + 1) % std::size(s_box);
        b = (b + s_box[a]) % std::size(s_box);

        // swap
        auto tmp = s_box[a];
        s_box[a] = s_box[b];
        s_box[b] = tmp;

        // crypt
        data[pos] ^= s_box[(s_box[a] + s_box[b]) % std::size(s_box)];
    }
}
